File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div1568-minimum-number-of-days-to-disconnect-island/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div1568-minimum-number-of-days-to-disconnect-island.py


class Solution:
    def minDays(self, grid: List[List[int]]) -> int:
        n,m=len(grid),len(grid[0])
        dire =[(0,1),(1,0),(-1,0),(0,-1)]
        
        def dfs(i,j,vis):
            
            vis[i][j]=1
            for dx,dy in dire:
                dx+=i
                dy+=j
            
                
                if 0<=dx<n and 0<=dy<m and not vis[dx][dy] and grid[dx][dy]:
                    dfs(dx,dy,vis)
                        
        
        
        def cntIslands(grid):
            islands=0
            vis = [[0]*(m) for _ in range(n)]
            for i in range(n):
                for j in range(m):
                    if not vis[i][j] and grid[i][j]:
                        dfs(i,j,vis)
                        islands+=1
            return islands
                    
        
        number_islands = cntIslands(grid)
        if number_islands>1 or number_islands==0:
            return 0
        
        
        for i in range(n):
            for j in range(m):
                if grid[i][j]:
                    grid[i][j]=0
                    number_islands= cntIslands(grid)
                    grid[i][j]=1
                    
                    if number_islands>1 or number_islands==0:
                        return 1
        return 2

File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div1568-minimum-number-of-days-to-disconnect-island/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-polygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div1568-minimum-number-of-days-to-disconnect-island.cpp
class Solution {
public:
    vector<int> dx={-1,0,1,0};
    vector<int> dy={0,-1,0,1};
    void dfs(int x, int y,   vector<vector<int>> & vis,   vector<vector<int>> &grid){
        int n = grid.size();
        int m = grid[0].size();

        vis[x][y]=1;
        
        for (int a=0;a<4;a++){
            int nx = x+dx[a];
            int ny = y+dy[a];
            
            if (0<=nx and nx<n and 0<=ny and ny<m and !vis[nx][ny] and grid[nx][ny]){
                dfs(nx,ny,vis,grid);
            }
        }
    }
    
    int cntIslands(vector<vector<int>> & grid){
        int islands =0;
        int n= grid.size();
        int m= grid[0].size();
        
        vector<vector<int>> vis(n,vector<int> (m,0));
        
        for (int i=0;i<n;i++)
        {
            for (int j=0;j<m;j++)
            {
                if (!vis[i][j] and grid[i][j])
                {
                    dfs(i,j,vis,grid);
                    islands++;
                }
            }
        }
        return islands;
    }
    
    int minDays(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        
        int num_islands = cntIslands(grid);
        //check for 0 ans
        if (num_islands>1 || num_islands==0)
        {
            return 0;
        }
        else
        {   // check for 1 ans
            for (int i=0; i<n; i++)
            {
                for (int j=0;j<m;j++)
                {
                    if (grid[i][j])
                    {
                        grid[i][j]=0;
                        num_islands = cntIslands(grid);
                        grid[i][j]=1;
                    
                        if (num_islands>1 or num_islands==0) 
                            return 1;
                    }
                }
            }
        }
        return 2;
    
        
    }
};